### PR TITLE
Update Barlow Twins with TF Similarity BarlowLoss function

### PR DESCRIPTION
**Pull request by dewball345**
This update replaces the previous loss function implementation with the one from TF Similarity. This is a drop-in replacement and the results look to be the same. The previous implementation is included for understanding purposes though it is preferable to use the updated one. Also changed the accuracy metrics because it was mentioned in the notebook that it was 64% though in the demo it went as high as 67%.

@ebursztein I could not find a way to use the augmenters with my program(current augmenters are keras.io layers which run probabilistically). I'm also a little concerned with `tf.map_func`, because I found that it slowed things down a lot.

### CODE REVIEW

**Comment by dewball345** [comment]
> This update replaces the previous loss function implementation with the one from TF Similarity. This is a drop-in replacement and the results look to be the same. The previous implementation is included for understanding purposes though it is preferable to use the updated one. Also changed the accuracy metrics because it was mentioned in the notebook that it was 64% though in the demo it went as high as 67%.

By the way, have not run `autogen.py`, but will do so after the python file has no errors to it.

Also, on a side note, I'm wondering why the Barlow Twin's title is being repeated twice on the website.

**Comment by owenvallis** [comment]
Hi @dewball345, I shared some timers for the tf.map_fn call in the tf.sim PR. Could you share some more context around the slowdown you were seeing? Thanks.

**Comment by dewball345** [comment]
> Hi @dewball345, I shared some timers for the tf.map_fn call in the tf.sim PR. Could you share some more context around the slowdown you were seeing? Thanks.

Thanks for the timers; i did not enclose my tf.map_fn in a @tf.function call. This was because when i did, i got an error about multiple tf.Variables being declared on a single function call, and this is likely attributed to the fact that I am structuring my augmentation layers as tf.keras layers.

**Comment by owenvallis** [comment]
I see, thanks for the info and I'll look into the tf.Variables error. In general, we've been adding the augmentation in the Dataset or Sampler objects and moving away from adding the augmentation as a model layer. There were a few reasons for this:
* We likely don't want to include the augmentation at inference time or in our validation and test data.
* Adding the augmentation as a model layer can cause some issues when training on TPU

**Comment by dewball345** [comment]
> I see, thanks for the info and I'll look into the tf.Variables error. In general, we've been adding the augmentation in the Dataset or Sampler objects and moving away from adding the augmentation as a model layer. There were a few reasons for this:
> 
> * We likely don't want to include the augmentation at inference time or in our validation and test data.
> * Adding the augmentation as a model layer can cause some issues when training on TPU

I see, that makes sense.

File: examples/vision/barlow_twins.py [hunk 1]
```diff
@@ -1,10 +1,11 @@
 """
 Title: Barlow Twins for Contrastive SSL
-Author: [Abhiraam Eranti](https://github.com/dewball345)
-Date created: 11/4/21
-Last modified: 12/20/21
+Author: [Abhiraam Eranti](https://github.com/dewball345)<br>
```
Comment (fchollet) [inline examples/vision/barlow_twins.py:3]: Remove br tags

File: examples/vision/barlow_twins.py [hunk 2]
```diff
@@ -126,15 +125,15 @@
...
-
 """
 
 """
 ## Setup
 """
 
 """shell
-pip install tensorflow-addons
+!pip install tensorflow-addons
```
Comment (fchollet) [inline examples/vision/barlow_twins.py:135]: Remove "!", they get added automatically

File: examples/vision/barlow_twins.py [hunk 3]
```diff
@@ -1037,12 +1047,15 @@ def train_step(self, batch: tf.Tensor) -> tf.Tensor:
 
 * [Paper](https://arxiv.org/abs/2103.03230)
 * [Original Pytorch Implementation](https://github.com/facebookresearch/barlowtwins)
-* [Sayak Paul's Implementation](https://colab.research.google.com/github/sayakpaul/Barlow-Twins-TF/blob/main/Barlow_Twins.ipynb#scrollTo=GlWepkM8_prl).
+* [Sayak Paul's
+Implementation](https://colab.research.google.com/github/sayakpaul/Barlow-Twins-TF/blob/main/Barlow_Twins.ipynb#scrollTo=GlWepkM8_prl).
```
Comment (fchollet) [inline examples/vision/barlow_twins.py:1051]: Make sure not to break markdown links, they should be on a single line each

**Comment by dewball345** [comment]
@fchollet Have done this.

File: examples/vision/barlow_twins.py [hunk 4]
```diff
@@ -604,10 +604,16 @@ def plot_values(batch: tuple):
 
 After this the two parts are summed together.
 
+We will be using the [BarlowLoss](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/lo
+sses/barlow.py)
```
Comment (fchollet) [inline examples/vision/barlow_twins.py:608]: Make sure all markdown links are unbroken

File: examples/vision/barlow_twins.py [hunk 5]
```diff
@@ -959,7 +967,9 @@ def train_step(self, batch: tf.Tensor) -> tf.Tensor:
 # chose the LAMB optimizer due to high batch sizes. Converged MUCH faster
 # than ADAM or SGD
 optimizer = tfa.optimizers.LAMB()
-loss = BarlowLoss(BATCH_SIZE)
+
+# We can just drop in either one of the two(results will be similar for both)
+loss = tensorflow_similarity.losses.Barlow()  # BarlowLoss(BATCH_SIZE)
```
Comment (fchollet) [inline examples/vision/barlow_twins.py:972]: What is the purpose of the comment `# BarlowLoss(BATCH_SIZE)`?

Comment (dewball345) [reply examples/vision/barlow_twins.py:972]: So that people can just drop in the from-scratch version and switch between the two.

Comment (dewball345) [reply examples/vision/barlow_twins.py:972]: (sorry i think i responded somewhere else but i'm not seeing it)

So that people can switch between the two easily.

File: examples/vision/barlow_twins.py (hunk 4 above)
Comment (dewball345) [reply examples/vision/barlow_twins.py:608]: I've checked the links and all of them do seem to work. Are you having any problems with some of the links?

**Review by fchollet — COMMENTED** [review]
Thank you! Are the generated files up to date?

**Comment by dewball345** [comment]
@fchollet No, not yet. I'm going to give that to you soon.

**Comment by pcoet** [comment]
Thanks for the PR, @dewball345. Are you still planning to generate the markdown and notebook files? Let us know if you're still working on this. Otherwise we'll close the request. Thanks!
